### PR TITLE
GasGadget for GAS opcode

### PR DESCRIPTION
**Pull request by roynalnaruto**
Implement the `GasGadget` for opcode `GAS` (5A).

Specs [here](https://github.com/scroll-tech/zkevm-specs/pull/25).

Other changes:
1. Fix constant gas cost for `OpcodeId::EXP` (comments below)

### CODE REVIEW

File: zkevm-circuits/src/evm_circuit/execution/gas.rs [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::GAS;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let gas_left = array_init(|_| cb.query_cell());
+
+        cb.require_equal(
+            "Constraint: gas left equal to stack value",
+            from_bytes::expr(&gas_left),
+            cb.curr.state.gas_left.expr() - 2.expr(),
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/execution/gas.rs:39]:
```suggestion
            cb.curr.state.gas_left.expr() - GasCost::QUICK.expr(),
```

**Review by icemelon — COMMENTED** [review]
I think you need to add GAS implementation in the bus-mapping (`bus-mapping/src/evm/opcodes/`) as well to pass the CI

File: bus-mapping/src/evm/opcodes/ids.rs [hunk 2]
```diff
@@ -500,7 +500,7 @@ impl OpcodeId {
             OpcodeId::SMOD => GasCost::FAST,
             OpcodeId::ADDMOD => GasCost::MID,
             OpcodeId::MULMOD => GasCost::MID,
-            OpcodeId::EXP => GasCost::ZERO,
+            OpcodeId::EXP => GasCost::SLOW,
```
Comment (roynalnaruto) [inline bus-mapping/src/evm/opcodes/ids.rs:503]: @icemelon The constant gas cost for `EXP` opcode is 10
Reference: [1](https://evm.codes) and [2](https://github.com/ethereum/solidity/blob/c61610302aa2bfa029715b534719d25fe3949059/libevmasm/GasMeter.h#L49)

**Review by roynalnaruto — COMMENTED** [review]
@icemelon added a comment for the gas cost bug fix for `OpcodeId::EXP`

**Comment by lispc** [comment]
Is it possible to use `StackOnlyOpcode::<0>::gen_associated_ops` like https://github.com/scroll-tech/zkevm-circuits/pull/73  ?

**Comment by roynalnaruto** [comment]
> Is it possible to use `StackOnlyOpcode::<0>::gen_associated_ops` like #73 ?

@lispc Yes its possible to use after the change: `step.stack.last_filled().map(|a| a - 1 + N)`

I had earlier seen `step.stack.nth_last_filled(N - 1)` hence decided to create a separate `Opcode` implementation. But I can re-use the most updated one (although the `TIMESTAMP` PR will need to be merged before then).

**Comment by lispc** [comment]
ok i suggest StackOnlyOpcode::<0> after merging of timestamp

File: zkevm-circuits/src/evm_circuit/param.rs [hunk 3]
```diff
@@ -32,3 +32,5 @@ pub(crate) const STACK_CAPACITY: usize = 1024;
 // counter to u64 as go-ethereum in case transaction size is allowed larger in
 // the future.
 pub(crate) const N_BYTES_PROGRAM_COUNTER: usize = N_BYTES_U64;
+pub(crate) const NUM_BYTES_PROGRAM_COUNTER: usize = N_BYTES_U64;
+pub(crate) const NUM_BYTES_GAS: usize = N_BYTES_U64;
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/param.rs:36]:
```suggestion

pub(crate) const N_BYTES_GAS: usize = N_BYTES_U64;
```

File: zkevm-circuits/src/evm_circuit/execution/gas.rs [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+};
+
+use array_init::array_init;
+use bus_mapping::evm::OpcodeId;
+use halo2::{arithmetic::FieldExt, circuit::Region, plonk::Error};
+
+#[derive(Clone, Debug)]
+pub(crate) struct GasGadget<F> {
+    same_context: SameContextGadget<F>,
+    value: RandomLinearCombination<F, NUM_BYTES_GAS>,
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/execution/gas.rs:25]:
```suggestion
    gas_left: RandomLinearCombination<F, NUM_BYTES_GAS>,
```

File: zkevm-circuits/src/evm_circuit/execution/gas.rs [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+impl<F: FieldExt> ExecutionGadget<F> for GasGadget<F> {
+    const NAME: &'static str = "GAS";
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::GAS;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        // The gas passed to a transaction is a 64-bit number.
+        let gas_left = array_init(|_| cb.query_cell());
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/execution/gas.rs:35]:
```suggestion
        let gas_left = cb.query_rlc();
```

File: zkevm-circuits/src/evm_circuit/execution/gas.rs [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+        // used by the `GAS` opcode itself.
+        cb.require_equal(
+            "Constraint: gas left equal to stack value",
+            from_bytes::expr(&gas_left),
+            cb.curr.state.gas_left.expr()
+                - OpcodeId::GAS.constant_gas_cost().expr(),
+        );
+
+        // Construct the value and push it to stack.
+        let value =
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/execution/gas.rs:47]: No need for `value` here.

File: zkevm-circuits/src/evm_circuit/execution/gas.rs [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+            RandomLinearCombination::new(gas_left, cb.power_of_randomness());
+        cb.stack_push(value.expr());
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/execution/gas.rs:49]:
```suggestion
        cb.stack_push(gas_left.expr());
```

File: zkevm-circuits/src/evm_circuit/execution/gas.rs [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+        let same_context = SameContextGadget::construct(
+            cb,
+            opcode,
+            step_state_transition,
+            None,
+        );
+
+        Self {
+            same_context,
+            value,
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/execution/gas.rs:67]:
```suggestion
            gas_left,
```

**Review by icemelon — APPROVED** [review]
LGTM. need to rebase

**Comment by icemelon** [comment]
Close and move to https://github.com/appliedzkp/zkevm-circuits/pull/301

**Pull request closed without merging**